### PR TITLE
refactor: model identifiers improvements

### PR DESCRIPTION
**Pull request by psychedelicious**
## What type of PR is this? (check all applicable)

- [x] Refactor
- [ ] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update
- [ ] Community Node Submission

## Have you discussed this change with the InvokeAI team?
- [x] Yes
- [ ] No, because:

## Description

- All models are identified by a key and optionally a submodel type via new model `ModelField`. Previously, a few model types had their own class, but not all of them. This inconsistency just added complexity without any benefit.
- Update all invocation to use the new format.
- In the node API, models are loaded by key or an instance of `ModelField` as a convenience.
- Add an enriched model schema for metadata. It includes key, hash, name, base and type.

Though the hash is recorded in metadata, it is not used to load models. The client will handle determining the model key for a model with fallback lookup via hash or model attributes. This functionality is not included in this PR, will be a followup (maybe v4.1.0).

With the changes to a lot of the node fields, I also fixed some spots where we had inconsistent capitalization.

## QA Instructions, Screenshots, Recordings

Everything should still work in the app, including metadata recall.

## Merge Plan
This PR can be merged when approved

## Added/updated tests?

- [x] Yes
- [ ] No : _please replace this line with details on why tests
      have not been included_

### CODE REVIEW

File: invokeai/frontend/web/src/features/nodes/util/graph/addT2IAdapterToLinearGraph.ts [hunk 1]
```diff
@@ -77,9 +81,18 @@ export const addT2IAdaptersToLinearGraph = (state: RootState, graph: NonNullable
         return;
       }
 
-      graph.nodes[t2iAdapterNode.id] = t2iAdapterNode as T2IAdapterInvocation;
+      graph.nodes[t2iAdapterNode.id] = t2iAdapterNode;
 
-      t2iAdapterMetdata.push(omit(t2iAdapterNode, ['id', 'type', 'is_intermediate']) as T2IAdapterField);
+      const modelConfig = await fetchModelConfigWithTypeGuard(t2iAdapter.model.key, isT2IAdapterModelConfig);
+
+      t2iAdapterMetdata.push({
```
Comment (maryhipp) [inline invokeai/frontend/web/src/features/nodes/util/graph/addT2IAdapterToLinearGraph.ts:88]:
```suggestion
      t2iAdapterMetadata.push({
```

File: invokeai/app/services/shared/invocation_context.py [hunk 2]
```diff
@@ -324,9 +328,13 @@ def load(self, key: str, submodel_type: Optional[SubModelType] = None) -> Loaded
...
 
-        return self._services.model_manager.load_model_by_key(
-            key=key, submodel_type=submodel_type, context_data=self._data
-        )
+        if isinstance(identifier, str):
+            model = self._services.model_manager.store.get_model(identifier)
+            return self._services.model_manager.load.load_model(model, submodel_type, self._data)
+        else:
+            _submodel_type = submodel_type or identifier.submodel_type
+            model = self._services.model_manager.store.get_model(identifier.key)
```
Comment (brandonrising) [inline invokeai/app/services/shared/invocation_context.py:336]: Is this were we were considering interpreting the key as either a hash or an ID? Maybe something like this?

```
  model: AnyModelConfig
  key = identifier if isinstance(identifier, str) else identifier.key
  submodel_type = submodel_type\
      if not (isinstance(identifier, ModelField) and identifier.submodel_type)\
      else identifier.submodel_type
  try:
    model = self._services.model_manager.store.get_model(key)
  except UnknownModelException:
    model = self._services.model_manager.store. get_model_by_hash(key)
  return self._services.model_manager.load.load_model(model, submodel_type, self._data)
```

Comment (psychedelicious) [reply invokeai/app/services/shared/invocation_context.py:336]: No, I realized we don't need to load models by hash. The client will handle reconciling metadata and retrieving the key for a model. Then the backend only loads by key.

**Review by maryhipp — APPROVED** [review]

**Review by hipsterusername — APPROVED** [review]

File: invokeai/app/services/shared/invocation_context.py (hunk 2 above)
Comment (brandonrising) [reply invokeai/app/services/shared/invocation_context.py:336]: I suppose not 🤔 I guess not a lot of people will interface with the API outside of the UI

**Pull request merged**